src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,431 @@
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return attnWt; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return attnWt; }
+
+  //! Get the value of deterministic.
+  OutputDataType const& Deterministic() const { return deterministic; }
+  //! Modify the value of deterministic.
+  OutputDataType& Deterministic() { return deterministic; }
+
+ private:
+  //! Locally-stored value of target sequence length.
+  size_t tgtSeqLen;
+
+  //! Locally-stored value of source sequence length.
+  size_t srcSeqLen;
+
+  //! Locally-stored module output size.
+  size_t embedDim;
+
+  //! Locally-stored number of parallel attention heads.
+  size_t numHeads;
+
+  //! Dimensionality of each head.
+  size_t headDim;
+
+  //! Locally-stored dropout rate used on output weights.
+  ElemType dropout;
+
+  //! Whether the forward pass is deterministic.
+  bool deterministic;
+
+  //! Two dimensional Attention Mask.
+  InputDataType attnMask2d;
+
+  //! Three dimensional Attention Mask.
+  arma::Cube<ElemType> attnMask3d;
+
+  //! Key Padding Mask.
+  InputDataType keyPaddingMask;
+
+  //! Locally-stored module output parameter visitor.
+  OutputDataType outputParameterVisitor;
+
+  //! Locally-stored module delta parameter visitor.
+  OutputDataType deltaVisitor;
+
+  //! Locally-stored attention output weight to be fed to last linear layer.
+  OutputDataType attnWt;
+
+  //! Locally-stored weight matrix associated with query.
+  OutputDataType queryWt;
+
+  //! Locally-stored weight matrix associated with key.
+  OutputDataType keyWt;
+
+  //! Locally-stored weight matrix associated with value.
+  OutputDataType valueWt;
+
+  //! Locally-stored weight matrix associated with attnWt.
+  OutputDataType outWt;
+
+  //! Locally-stored bias associated with query.
+  OutputDataType queryBias;
+
+  //! Locally-stored bias associated with key.
+  OutputDataType keyBias;
+
+  //! Locall-stored bias associated with value.
+  OutputDataType valueBias;
+
+  //! Locally-stored bias associated with attnWt.
+  OutputDataType outBias;
+
+  //! Softmax layer to represent the probabilities of next sequence.
+  Softmax<InputDataType, OutputDataType>* softmaxModule;
+
+  //! Dropout layer (optional).
+  Dropout<InputDataType, OutputDataType>* dropoutModule;
+
+  //! Locally-stored delta object.
+  OutputDataType delta;
+
+  //! Locally-stored gradient object.
+  OutputDataType gradient;
+
+  //! Locally-stored output parameter.
+  OutputDataType outputParameter;
+}; // class MultiheadAttention
+
+/**
+ * Self Attention Policy class. It deals with the constraints of the parameters
+ * of the MultiheadAttention class.
+ */
+class SelfAttention
+{
```

Comment:
@lozhnikov: I am wondering where should I put these policy classes? For now, I've kept them in `multihead_attention.hpp` header file.